

*Multiple Linear Regression Model*
*for Determining Highway mileage*

In [ ]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Load Dataset
df = pd.read_csv("cars.csv")



Load the dataset

In [ ]:
df.head()

Dimensions.Height  Dimensions.Length  Dimensions.Width  \
0                140                143               202   
1                140                143               202   
2                140                143               202   
3                140                143               202   
4                140                143               202   

  Engine Information.Driveline                Engine Information.Engine Type  \
0              All-wheel drive          Audi 3.2L 6 cylinder 250hp 236ft-lbs   
1            Front-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
2            Front-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
3              All-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   
4              All-wheel drive  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo   

   Engine Information.Hybrid  Engine Information.Number of Forward Gears  \
0                       True                                           6   
1                       True                                           6   
2                       True                                           6   
3                       True                                           6   
4                       True                                           6   

  Engine Information.Transmission  Fuel Information.City mpg  \
0  6 Speed Automatic Select Shift                         18   
1  6 Speed Automatic Select Shift                         22   
2                  6 Speed Manual                         21   
3  6 Speed Automatic Select Shift                         21   
4  6 Speed Automatic Select Shift                         21   

  Fuel Information.Fuel Type  Fuel Information.Highway mpg  \
0                   Gasoline                            25   
1                   Gasoline                            28   
2                   Gasoline                            30   
3                   Gasoline                            28   
4                   Gasoline                            28   

  Identification.Classification           Identification.ID  \
0        Automatic transmission            2009 Audi A3 3.2   
1        Automatic transmission       2009 Audi A3 2.0 T AT   
2           Manual transmission          2009 Audi A3 2.0 T   
3        Automatic transmission  2009 Audi A3 2.0 T Quattro   
4        Automatic transmission  2009 Audi A3 2.0 T Quattro   

  Identification.Make Identification.Model Year  Identification.Year  \
0                Audi              2009 Audi A3                 2009   
1                Audi              2009 Audi A3                 2009   
2                Audi              2009 Audi A3                 2009   
3                Audi              2009 Audi A3                 2009   
4                Audi              2009 Audi A3                 2009   

   Engine Information.Engine Statistics.Horsepower  \
0                                              250   
1                                              200   
2                                              200   
3                                              200   
4                                              200   

   Engine Information.Engine Statistics.Torque  
0                                          236  
1                                          207  
2                                          207  
3                                          207  
4                                          207

Rename for colunms better readebility

In [ ]:
df.rename(columns={
    'Dimensions.Height': 'Height',
    'Dimensions.Length': 'Length',
    'Dimensions.Width': 'Width',
    'Engine Information.Driveline': 'Driveline',
    'Engine Information.Engine Type': 'Engine_Type',
    'Engine Information.Hybrid': 'Hybrid',
    'Engine Information.Number of Forward Gears': 'Forward_Gears',
    'Engine Information.Transmission': 'Transmission',
    'Fuel Information.City mpg': 'City_MPG',
    'Fuel Information.Fuel Type': 'Fuel_Type',
    'Fuel Information.Highway mpg': 'Highway_MPG',
    'Identification.Classification': 'Classification',
    'Identification.ID': 'ID',
    'Identification.Make': 'Make',
    'Identification.Model Year': 'Model_Year',
    'Identification.Year': 'Year',
    'Engine Information.Engine Statistics.Horsepower': 'Horsepower',
    'Engine Information.Engine Statistics.Torque': 'Torque'
}, inplace=True)

df.head()

Height  Length  Width          Driveline  \
0     140     143    202    All-wheel drive   
1     140     143    202  Front-wheel drive   
2     140     143    202  Front-wheel drive   
3     140     143    202    All-wheel drive   
4     140     143    202    All-wheel drive   

                                    Engine_Type  Hybrid  Forward_Gears  \
0          Audi 3.2L 6 cylinder 250hp 236ft-lbs    True              6   
1  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo    True              6   
2  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo    True              6   
3  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo    True              6   
4  Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo    True              6   

                     Transmission  City_MPG Fuel_Type  Highway_MPG  \
0  6 Speed Automatic Select Shift        18  Gasoline           25   
1  6 Speed Automatic Select Shift        22  Gasoline           28   
2                  6 Speed Manual        21  Gasoline           30   
3  6 Speed Automatic Select Shift        21  Gasoline           28   
4  6 Speed Automatic Select Shift        21  Gasoline           28   

           Classification                          ID  Make    Model_Year  \
0  Automatic transmission            2009 Audi A3 3.2  Audi  2009 Audi A3   
1  Automatic transmission       2009 Audi A3 2.0 T AT  Audi  2009 Audi A3   
2     Manual transmission          2009 Audi A3 2.0 T  Audi  2009 Audi A3   
3  Automatic transmission  2009 Audi A3 2.0 T Quattro  Audi  2009 Audi A3   
4  Automatic transmission  2009 Audi A3 2.0 T Quattro  Audi  2009 Audi A3   

   Year  Horsepower  Torque  
0  2009         250     236  
1  2009         200     207  
2  2009         200     207  
3  2009         200     207  
4  2009         200     207

Removing unnecessary variables


In [ ]:
# Keep only relevant columns
df = df[['Height', 'Length', 'Width', 'Driveline', 'Hybrid', 'Forward_Gears',
         'City_MPG', 'Fuel_Type', 'Highway_MPG', 'Horsepower', 'Torque']]

df.head()

Height  Length  Width          Driveline  Hybrid  Forward_Gears  City_MPG  \
0     140     143    202    All-wheel drive    True              6        18   
1     140     143    202  Front-wheel drive    True              6        22   
2     140     143    202  Front-wheel drive    True              6        21   
3     140     143    202    All-wheel drive    True              6        21   
4     140     143    202    All-wheel drive    True              6        21   

  Fuel_Type  Highway_MPG  Horsepower  Torque  
0  Gasoline           25         250     236  
1  Gasoline           28         200     207  
2  Gasoline           30         200     207  
3  Gasoline           28         200     207  
4  Gasoline           28         200     207

Remove Duplicate Values

In [ ]:
#Droping Duplicates
df=df.drop_duplicates()

One-hot Encoding

In [ ]:
# One-hot encode Driveline and Fuel_Type
df = pd.get_dummies(df, columns=['Driveline', 'Fuel_Type'], drop_first=True)



# Convert categorical variables into numerical (One-Hot Encoding)
df = pd.get_dummies(df, drop_first=True)


In [ ]:
df['Hybrid'] = df['Hybrid'].astype(int)
df['Driveline_Four-wheel drive'] = df['Driveline_Four-wheel drive'].astype(int)
df['Driveline_Front-wheel drive'] = df['Driveline_Front-wheel drive'].astype(int)
df['Driveline_Rear-wheel drive'] = df['Driveline_Rear-wheel drive'].astype(int)
df['Fuel_Type_Diesel fuel'] = df['Fuel_Type_Diesel fuel'].astype(int)
df['Fuel_Type_E85'] = df['Fuel_Type_E85'].astype(int)
df['Fuel_Type_Gasoline'] = df['Fuel_Type_Gasoline'].astype(int)

In [ ]:
# Set 'Horsepower' as the target variable

X = df.drop(columns=['Horsepower'])  # Independent variables
y = df['Horsepower']  # Target variable

VIF Score

In [ ]:
# Add constant for VIF calculation
X = sm.add_constant(X)

# VIF Calculation
vif_data = pd.DataFrame()
vif_data['Feature'] = X.columns
vif_data['VIF Score'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

# Display VIF Scores
print(vif_data)


                        Feature    VIF Score
0                        Height     1.137723
1                        Length     1.016085
2                         Width     1.155196
3                        Hybrid  2423.188189
4                 Forward_Gears     1.241679
5                      City_MPG     5.718378
6                   Highway_MPG     2.668250
7                        Torque     2.926681
8    Driveline_Four-wheel drive     1.699005
9   Driveline_Front-wheel drive     2.479182
10   Driveline_Rear-wheel drive     1.949344
11        Fuel_Type_Diesel fuel    16.982080
12                Fuel_Type_E85   206.276980
13           Fuel_Type_Gasoline   218.247666


Feature Selection, Model Training and evaluvation

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split # Import train_test_split
from sklearn.linear_model import LinearRegression # Import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Select the final independent variables features = ['Height', 'Length', 'Width', 'Forward_Gears', 'Torque',             'Driveline_Four-wheel drive', 'Driveline_Front-wheel drive', 'Driveline_Rear-wheel drive']

features = ['Horsepower','Length','Width','Highway_MPG']

X = df[features]
y = df['City_MPG']


# Splitting into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creating and training the MLR model
model = LinearRegression()
model.fit(X_train, y_train)

# Making predictions on the test set
y_pred = model.predict(X_test)

# Evaluating the model
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

# Display model performance
print(f"R² Score: {r2:.4f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")

# Display feature coefficients
coefficients = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_})
print("\nModel Coefficients:")
print(coefficients)



R² Score: 0.8225
Mean Absolute Error (MAE): 1.48
Mean Squared Error (MSE): 3.78
Root Mean Squared Error (RMSE): 1.94

Model Coefficients:
       Feature  Coefficient
0   Horsepower    -0.019339
1       Length    -0.001229
2        Width    -0.003239
3  Highway_MPG     0.327300


Outlier Removel

In [ ]:
# Calculate the IQR for each numeric column
Q1 = df.select_dtypes(include=[np.number]).quantile(0.25)
Q3 = df.select_dtypes(include=[np.number]).quantile(0.75)
IQR = Q3 - Q1

# Define the lower and upper bounds for outliers
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter the DataFrame to keep only the rows within the bounds
df_no_outliers = df[~((df.select_dtypes(include=[np.number]) < lower_bound) | (df.select_dtypes(include=[np.number]) > upper_bound)).any(axis=1)]

# Display the new DataFrame without outliers
print(df_no_outliers)


      Height  Length  Width  Hybrid  Forward_Gears  City_MPG  Highway_MPG  \
0        140     143    202       1              6        18           25   
1        140     143    202       1              6        22           28   
2        140     143    202       1              6        21           30   
3        140     143    202       1              6        21           28   
5         91      17     62       1              6        16           27   
...      ...     ...    ...     ...            ...       ...          ...   
5044      28      64    135       1              6        17           25   
5046      28      64    135       1              6        16           23   
5049     175     153    138       1              6        17           24   
5067       3     253    201       1              5        18           25   
5071      13     253    201       1              5        18           25   

      Horsepower  Torque  Driveline_Four-wheel drive  \
0            250   

Model Training After Removing Outliers

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split # Import train_test_split
from sklearn.linear_model import LinearRegression # Import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

# Select the final independent variables features = ['Height', 'Length', 'Width', 'Forward_Gears', 'Torque',             'Driveline_Four-wheel drive', 'Driveline_Front-wheel drive', 'Driveline_Rear-wheel drive']

features = ['Horsepower','Length','Width','Highway_MPG']

X = df_no_outliers[features]
y = df_no_outliers['City_MPG']


# Splitting into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Creating and training the MLR model
model = LinearRegression()
model.fit(X_train, y_train)

# Making predictions on the test set
y_pred = model.predict(X_test)

# Evaluating the model
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

# Display model performance
print(f"R² Score: {r2:.4f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")

# Display feature coefficients
coefficients = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_})
print("\nModel Coefficients:")
print(coefficients)



R² Score: 0.9037
Mean Absolute Error (MAE): 0.93
Mean Squared Error (MSE): 1.42
Root Mean Squared Error (RMSE): 1.19

Model Coefficients:
       Feature  Coefficient
0   Horsepower    -0.011568
1       Length    -0.000706
2        Width     0.002813
3  Highway_MPG     0.606767


**Evaluation before removing outliers**

1. R² Score: 0.8225
2. Mean Absolute Error (MAE): 1.48
3. Mean Squared Error (MSE): 3.78
4. Root Mean Squared Error (RMSE): 1.94

**Evaluation after removing outliers**

1. R² Score: 0.9037
2. Mean Absolute Error (MAE): 0.93
3. Mean Squared Error (MSE): 1.42
4. Root Mean Squared Error (RMSE): 1.19

